groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+            this->d_type = other.d_type;
+        }
+        other.reset();
+    }
+}
+
+// ACCESSORS
+
+template <class HEAD, class... TAIL>
+inline
+bool variant<HEAD, TAIL...>::valueless_by_exception()
+const BSLS_KEYWORD_NOEXCEPT
+{
+    return d_type == bsl::variant_npos;
+}
+
+template <class HEAD, class... TAIL>
+inline
+size_t variant<HEAD, TAIL...>::index() const BSLS_KEYWORD_NOEXCEPT
+{
+    return d_type;
+}
+
+template <class HEAD, class... TAIL>
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+template <bool, typename>
+#endif  //  BSL_VARIANT_FULL_IMPLEMENTATIO
+typename variant<HEAD, TAIL...>::allocator_type
+variant<HEAD, TAIL...>::get_allocator()
+const BSLS_KEYWORD_NOEXCEPT
+{
+    return allocator_type(AllocBase::mechanism());
+}
+
+// FREE functions
+
+template <class HEAD, class... TAIL>
+void swap(bsl::variant<HEAD, TAIL...>& lhs,
+          bsl::variant<HEAD, TAIL...>& rhs)  // noexcept(see below );
+{
+    BloombergLP::bslstl::VariantSwapImpl(
+        bsl::integral_constant<
+            bool,
+            BloombergLP::bslma::UsesBslmaAllocatorAny<HEAD, TAIL...>::value>(),
+        lhs,
+        rhs);
+}
+
+// HASH SPECIALIZATIONS
+
+template <class HASHALG, class HEAD, class... TAIL>
+void hashAppend(HASHALG& hashAlg, const variant<HEAD, TAIL...>& input)
+
+{
+    if (!input.valueless_by_exception()) {
+        hashAppend(hashAlg, input.index());
+        BloombergLP::bslstl::Variant_HashVisitor<HASHALG> hashVisitor(hashAlg);
+        visit(hashVisitor, input);
+    }
+    else {
+        hashAppend(hashAlg, false);
+    }
+}
+
+template <class TYPE, class HEAD, class... TAIL>
+bool holds_alternative(const variant<HEAD, TAIL...>& obj) BSLS_KEYWORD_NOEXCEPT
+{
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_STATIC_ASSERT
+    static_assert(
+        BloombergLP::bslstl::
+                Variant_HasUniqueType<TYPE, variant<HEAD, TAIL...> >::value ==
+            true,
+        "Type is not a unique alternative in the variant");
+#endif
+    typedef typename bsl::variant<HEAD, TAIL...> VARIANT;
+    return obj.index() == BSLSTL_VARIANT_INDEX_OF(TYPE, VARIANT);
+}
+
+template <size_t INDEX, class HEAD, class... TAIL>
+typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type& get(
+                                                   variant<HEAD, TAIL...>& obj)
+{
+    BSLMF_ASSERT(INDEX < (sizeof...(TAIL) + 1));
+
+    typedef typename bsl::variant<HEAD, TAIL...>                   VARIANT;
+    typedef typename BloombergLP::bslstl::Variant_ImpUtil<VARIANT> Var_ImpUtil;
+    typedef typename bsl::variant_alternative<INDEX, VARIANT>::type RETURNTYPE;
+
+    return Var_ImpUtil::template getImpl<RETURNTYPE>(
+        bsl::in_place_index_t<INDEX>(), obj);
+}
+
+template <size_t INDEX, class HEAD, class... TAIL>
+const typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type& get(
+                                             const variant<HEAD, TAIL...>& obj)
+{
+    BSLMF_ASSERT(INDEX < ((sizeof...(TAIL) + 1) + 1));
```

Comment:
leftover from old code. Was about to replace all of that +1 nonsense with `variant_size` but while adding tests i realised the BSLMF_ASSERT is never reached for invalid indices - the return type formation fails for such types. 
I suggest I remove all these asserts.
If we want better error messages, we can consider doing a static_assert(false) like trickery in the `variant_alternative`. Thoughts?

Reply:
Yeah I think it's a good idea to put a static assertion in the `variant_alternative` logic somewhere. I'll leave it to you to decide whether this is worth spending time thinking about right now (it's not immediately obvious to me where it would go).